api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+#include "opentelemetry/common/attribute_value.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/trace/key_value_iterable_view.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+// The context class provides a context identifier.
+// This is a dummy class that is meant to be overridden,
+// the methods return default values.
+class Context
+{
+
+public:
+  // The Key class is used to obscure access from the
+  //  user to the context map. The identifier is used as a key
+  // to the context map.
+  class Key
+  {
+
+  public:
+    // Returns the key's identifier.
+    nostd::string_view GetIdentifier() { return ""; }
+
+    // Returns the key's name
+    nostd::string_view GetName() { return ""; }
+
+    Key() = default;
+
+    // Constructs a new Key with the passed in name. Sets the identifier as
+    // the address of this object.
+    Key(nostd::string_view key_name) {}
+
+  private:
+    friend class Context;
+  };
+
+  // Creates a key with the passed in name and returns it.
+  Key CreateKey(nostd::string_view key_name) { return Key(key_name); }
+
+  Context() = default;
+
+  // Contructor, creates a context object from a map of keys
+  // and identifiers.
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context(const T &attributes)
+  {}
+
+  // Accepts a new iterable and then returns a new  context that
+  // contains both the original pairs and the new pair.
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context WriteValues(T &attributes) noexcept
+  {
+    return Context(attributes);
+  }
+
+  // Returns the value associated with the passed in key.
+  virtual common::AttributeValue GetValue(Key key) { return 0; }
```

Comment:
I'm doubtful whether `common::AttributeValue` will cover all use cases. One use case might be storing the current `Span` or `SpanContext`, and this is not covered by the `AttributeValue` variant.

Reply:
What might we return instead of common::AttributeValue?

Reply:
`opentelemetry/nostd/variant.h`

Reply:
I'm not completely sure.

One option might be to define a list of all value types we need to support and then add another context specific variant.

Reply:
Would it be valuable to have a struct for ```SpanContext``` and just allow it be in ```common::AttributeValue``` ? Or do you see value in having generic any key-value context be nested in a variant? I mean ```CorrelationContext``` is probably just a dictionary of existing ```nostd::variant``` as long as we add ```SpanContext``` to it?

Reply:
I think `AttributeValue` is designed for [`Span Attribute`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-attributes), it should be changed to cover `SpanContext`.

Reply:
> One use case might be storing the current Span or SpanContext

I pointed Sam at AttributeValue (on gitter) based on the assumption that spans would be by ID (not as an object) using a specific key in the current context, and when the span changes the new ID is stored for that key.

If it's not clear what the set of value types needs to be, can we start with
`  using ContextValue = AttributeValue;
`
until the issue is resolved (either be removing the type alias or defining ContextValue concretely?

Reply:
Can we make different assumptions about the context types:

- Context
- SpanContext

being really just structs, possibly with getters/setters.

And only:
- CorrelationContext

being a dictionary, in traditional C++ sense of it? (map or unordered_map, with extra validation of input constraints on key / naming convention of it, possibly throwing or terminating for invalid key names)

Reply:
I looked at how Java OT operates with Jaeger exporter, and how JaegerSpanContext is implemented: pretty much a "struct" (Plain Java Object), with helper getter / setters for input / output validation. Can we consider that some context types are more "struct"-alike, and the other is more "dictionary"-alike? We can spec common getter / setter pattern for these, if needed make them share common ancestor, but semantically the best type for a dictionary in C++ is likely `std::unordered_map<nostd::string_view, common::AtributeValue>` ... Would there be a concern if span event attributes and correlation context attributes be of the same common variant type?

Reply:
Who needs to be involved to make the decision on this?

Reply:
Let's discuss this topic on Monday community meeting.